File: leetcode/1584_Min_Cost_to_Connect_All_Points.py
from typing import List
from unittest import TestCase

class DisjointSet: 
    def __init__(self, n: int) -> None:
        self.parents = [i for i in range(n)]
        self.ranks = [0] * n

    def root(self, i: int) -> int:
        while i != self.parents[i]:
            i, self.parents[i] = self.parents[i], self.parents[self.parents[i]]
        return i

    def join(self, i: int, j: int) -> None:
        root_i, root_j = (self.root(i), self.root(j))
        if self.ranks[root_i] < self.ranks[root_j]:
            self.parents[root_i] = root_j
        elif self.ranks[root_i] > self.ranks[root_j]:
            self.parents[root_j] = root_i
        else:
            self.parents[root_i] = root_j
            self.ranks[root_j] += 1

class Solution(object):
    def minCostConnectPoints(self, points: List[List[int]]) -> int:
        n = len(points)
        ds, edges = DisjointSet(n), []
        total_cost, edge_cnt = 0, 0

        for i in range(n):
            for j in range(i + 1, n):
                edges.append((abs(points[i][0] - points[j][0]) + abs(points[i][1] - points[j][1]), i ,j))
        edges.sort()
        for cost, i, j in edges:
            if ds.root(i) != ds.root(j):
                ds.join(i, j)
                total_cost += cost
                edge_cnt += 1
                if edge_cnt == n - 1:
                    break
        return total_cost

class MinCostConnectPointsTest(TestCase):
    def test_1(self) -> None:
        self.assertEqual(20, Solution().minCostConnectPoints([[0,0],[2,2],[3,10],[5,2],[7,0]]))

    def test_2(self) -> None:
        self.assertEqual(18, Solution().minCostConnectPoints([[3,12],[-2,5],[-4,1]]))

File: leetcode/1584_Min_Cost_to_Connect_All_Points.rs
struct Solution {}

struct DisjointSet {
    parents: Vec<usize>,
    ranks: Vec<usize>,
}

impl DisjointSet {
    fn new(n: usize) -> Self {
        Self { parents: (0..n).collect(), ranks: vec![0; n] }
    }

    fn find(&mut self, mut i: usize) -> usize {
        while i != self.parents[i] {
            let tmp = i;
            i = self.parents[i];
            self.parents[tmp] = self.parents[i];
        }
        i
    }

    fn join(&mut self, i: usize, j: usize) {
        let (root_i, root_j) = (self.find(i), self.find(j));
        if self.ranks[root_i] < self.ranks[root_j] {
            self.parents[root_i] = root_j;
        } else if self.ranks[root_i] > self.ranks[root_j] {
            self.parents[root_j] = root_i;
        } else {
            self.parents[root_i] = root_j;
            self.ranks[root_j] += 1;
        }
    }
}

impl Solution {
    pub fn min_cost_connect_points(points: Vec<Vec<i32>>) -> i32 {
        let n = points.len();
        let mut edges = Vec::new();
        let mut ds = DisjointSet::new(n);
        let mut total_cost = 0;
        let mut edges_cnt = 0;
        for i in 0..n {
            for j in i + 1..n {
                edges.push(((points[i][0] - points[j][0]).abs() + (points[i][1] - points[j][1]).abs(), i, j))
            }
        }
        edges.sort();
        for (cost, i, j) in edges {
            if ds.find(i) != ds.find(j) {
                ds.join(i, j);
                total_cost += cost;
                edges_cnt += 1;
                if edges_cnt == n - 1 {
                    break;
                }
            }
        }
        total_cost
    }
}

#[test]
fn test_1() {
    assert_eq!(Solution::min_cost_connect_points(vec![vec![0,0], vec![2,2], vec![3,10], vec![5,2], vec![7,0]]), 20);
}

#[test]
fn test_2() {
    assert_eq!(Solution::min_cost_connect_points(vec![vec![3,12], vec![-2,5], vec![-4,1]]), 18);
}

